src/main/java/org/embulk/input/google_ads/GoogleAdsReporter.java

```diff
@@ -53,40 +57,44 @@ private UserCredentials buildCredential(PluginTask task)
                 .build();
     }
 
-    public Iterable<GoogleAdsServiceClient.SearchPage> getReportPage()
-    {
-        List<GoogleAdsServiceClient.SearchPage> pages = new ArrayList<GoogleAdsServiceClient.SearchPage>();
-
-        String startDateTime = null;
-       do {
-            String query = buildQuery(task, startDateTime);
-            logger.info(query);
-            SearchGoogleAdsRequest request = buildRequest(task, query);
-            GoogleAdsServiceClient googleAdsService = client.getVersion14().createGoogleAdsServiceClient();
-            GoogleAdsServiceClient.SearchPagedResponse response = googleAdsService.search(request);
-
-            if (response.getPage().getResponse().getResultsCount() == 0) {
-                return pages;
-            }
+    private Iterable<GoogleAdsServiceClient.SearchPage> search(Map<String, String> params) {
+        String query = buildQuery(task, params);
+        logger.info(query);
+        SearchGoogleAdsRequest request = buildRequest(task, query);
+        GoogleAdsServiceClient googleAdsService = client.getVersion14().createGoogleAdsServiceClient();
+        GoogleAdsServiceClient.SearchPagedResponse response = googleAdsService.search(request);
+        return response.iteratePages();
+    }
 
-            response.iteratePages().iterator().forEachRemaining(pages::add);
+    public void search(Consumer<Iterable<GoogleAdsServiceClient.SearchPage>> consumer, Map<String, String> params) {
```

Comment:
nits: Although Iterator is called twice to get the last row, it is called once when the search method params change to per page.

```ruby
public void search(Consumer<GoogleAdsServiceClient.SearchPage> consumer, Map<String, String> params) {
     Iterable<GoogleAdsServiceClient.SearchPage> pages = search(params);
    for (GoogleAdsServiceClient.SearchPage page : pages) {
        consumer.accept(page);
        lastPage = page;
    }

    if (!task.getResourceType().equals("change_event") || lastPage == null) {
        return;
     }

     ...

```